java-checks/src/main/java/org/sonar/java/checks/regex/ReluctantQuantifierCheck.java

```diff
@@ -104,9 +108,30 @@ private String makePossessive(Quantifier quantifier) {
       return Optional.of(result);
     }
 
+    private boolean hasNoIntersection(CharacterClassElementTree tree, @Nullable CharacterClassElementTree base) {
+      return base != null && !new SimplifiedRegexCharacterClass(base, new FlagSet()).intersects(new SimplifiedRegexCharacterClass(tree, new FlagSet()), false);
```

Comment:
Instead of `new FlagSet()`, we should be calling `getFlagSet()` here. Otherwise it won't take flags into account for cases like `"(?i)[a-z]*?X"` or `Pattern.compile("[a-z]*?X", Pattern.CASE_INSENSITIVE)`.

Even with that change we can still get wrong results if a flag is only enabled for part of the regex (e.g. `(?i:[a-z]*?)X` where the `i` flag would already be unset by the time we call `getFlagSet`), but that's actually pretty tricky to handle with the current API. @alban-auzeill  is currently working on a change that will make it easy to query the flags for a specific RegexTree, so it's best to just use `getFlags()` for now and then we can replace it with `base.getFlags()` and `tree.getFlags()` respectively afterwards once Alban adds that method.